Reduce global helper functions

distantnative opened the pull request:
## This PR …
<!--
A clear and concise description of the PR.
Use this section for review hints, explanations or discussion points/todos.

Add relevant release notes: Features, Enhancements, Fixes, Deprecated.
Reference issues from the `kirby` repo or ideas from `feedback.getkirby.com`.
Always mention whether your PR introduces breaking changes.

How to contribute: https://contribute.getkirby.com
-->

### Enhancements
- Helpers can be deactivated by setting a global constant named `KIRBY_HELPER_*`, e.g. `KIRBY_HELPER_DUMP` to `false`

  In some cases, our global helper functions would collide with other tools' global functions as they would use the same name, e.g. `dump()`. If you are affected by this problem, you can now set the matching constant, e.g. `KIRBY_HELPER_DUMP`, to `false` and Kirby will not register its global helper.

  To ensure that Kirby still functions properly, we moved all functionality from the helper functions to class methods (and the global helper functions are only aliases now). We recommend plugin developers to also start using these class methods to ensure that your code always uses Kirby's core helpers:

  - New `Cms\App::csrf()` method with the functionality of the `csrf()` helper
  - New `Cms\App::image()` method with the functionality of the `image()` helper
  - New `Cms\Helpers::deprecated()` method with the functionality of the `deprecated()` helper
  - New `Cms\Helpers::size()` method with the functionality of the `size()` helper
  - New `Cms\Helpers::dump()` method with the functionality of the `dump()` helper
  - New `Cms\Html::css()` method with the functionality of the `css()` helper
  - New `Cms\Html::js()` method with the functionality of the `js()` helper
  - New `Cms\Html::svg()` method with the functionality of the `svg()` helper
  - New `Filesystem\F::loadClasses()` method with the functionality of the `load()` helper
  - New `Http\Router::execute()` method with the functionality of the `router()` helper
  - New `Http\Response::go()` method with the functionality of the `go()` helper
  https://kirby.nolt.io/369
  - New `Toolkit\Date::roundedTimestamp()` method with the functionality of the `timestamp()` helper
  - New `Toolkit\Str::esc()` method with the functionality of the `esc()` helper
  - New `Toolkit\Str::uuid()` method with the functionality of the `uuid()` helper
  - New `Toolkit\V::invalid()` method with the functionality of the `invalid()` helper
  - Extended `Cms\App::kirbytag()` method with the functionality of the `kirbytag()` helper. Allows to pass an array as first parameter.
  - Extended `Cms\App::snippet()` method with the functionality of the `snippet()` helper.
  - Extended `Toolkit\Html::attr()` method with the functionality of the `attr()` helper. Allows to pass `$before` and `$after` string.

### Refactoring

- The Kirby core no longer uses Kirby's helper functions internally but calls the underlying methods directly.

### Deprecated

- The `dump` component was deprecated and superseded by the new feature to replace helpers (`KIRBY_HELPER_DUMP` constant). The component will be removed in Kirby 3.8.0.

### Breaking changes

- `Toolkit\Html::attr([], null)` now consistently returns `null` instead of an empty string.


## Related backlog issues

https://github.com/getkirby/backlog/issues/39
https://github.com/getkirby/backlog/issues/38
https://github.com/getkirby/backlog/issues/18

## Ready?
<!--
If you can help to check off the following tasks, that'd be great.
If not, don't worry - we will take care of it.

More details: https://contribute.getkirby.com
-->

- [x] Unit tests for fixed bug/feature
- [x] In-code documentation (wherever needed)
- [x] Tests and checks all pass


### For review team
<!-- 
We will take care of the following before merging the PR.
-->

- [x] Add changes to release notes draft in Notion
- [x] Add to [website docs release checklist](https://github.com/getkirby/getkirby.com/pulls) (if needed)

distantnative committed:
Bring back `App::image()`

distantnative committed:
Change unit test for breaking change

distantnative committed:
Fix since tags

distantnative committed:
Reduce global helper functions

distantnative committed:
Adapt helper methods naming

lukasbestle reviewed (commented):
I am thinking we could call `kirby()` instead of `App::instance()` in all the helpers that depend on the `App` instance (`collection()`, `kirbytext()`, `option()`...). This would allow users to override the single `kirby()` helper to change the `App` object for all headers at once.

lukasbestle commented on src/Cms/App.php:
```diff
@@ -529,6 +529,43 @@ public function core()
+     * @param string|null $check Pass a token here to compare it to the one in the session
+     * @return string|bool Either the token or a boolean check result
+     */
+    public static function csrf(?string $check = null)
```
We could make this an instance method to be in line with `session()` and all the others. Then we wouldn't need to use `static::instance()` inside but just `$this`.

lukasbestle commented on src/Cms/App.php:
```diff
@@ -770,14 +853,27 @@ public function io($input)
      *
      * @internal
-     * @param string $type
+     * @param string|array $type
```
```suggestion
     * @param string|array $type Tag type or array with all tag arguments
                                 (the key of the first element becomes the type)
```

lukasbestle commented on src/Cms/App.php:
```diff
@@ -1506,11 +1602,22 @@ public function smartypants(string $text = null): string
      * @internal
      * @param mixed $name
      * @param array $data
+     * @param bool $return
```
```suggestion
     * @param array|object $data Variables or an object that becomes `$item`
     * @param bool $return On `false`, directly echo the snippet
```

lukasbestle commented on src/Cms/App.php:
```diff
@@ -1506,11 +1602,22 @@ public function smartypants(string $text = null): string
      * @return string|null
      */
-    public function snippet($name, array $data = []): ?string
+    public function snippet($name, array $data = [], bool $return = true): ?string
```
```suggestion
    public function snippet($name, $data = [], bool $return = true): ?string
```

lukasbestle commented on src/Cms/Auth.php:
```diff
@@ -195,7 +195,7 @@ public function csrf()
     {
         $isDev = $this->kirby->option('panel.dev', false) !== false;
-        return $this->kirby->option('api.csrf', $isDev ? 'dev' : csrf());
+        return $this->kirby->option('api.csrf', $isDev ? 'dev' : App::csrf());
```
```suggestion
        return $this->kirby->option('api.csrf', $isDev ? 'dev' : $this->kirby->csrf());
```

lukasbestle commented on src/Cms/Html.php:
```diff
@@ -15,6 +18,50 @@
+    public static function css($url, $options = null): ?string
+    {
+        if (is_array($url) === true) {
+            $links = A::map($url, fn ($url) => css($url, $options));
```
```suggestion
            $links = A::map($url, fn ($url) => static::css($url, $options));
```

lukasbestle commented on src/Toolkit/Html.php:
```diff
@@ -133,9 +133,11 @@ public static function a(string $href, $text = null, array $attr = []): string
      *                           Key-value array: A list of attributes will be generated. Don't pass a second argument in that case.
      * @param mixed $value If used with a `$name` string, pass the value of the attribute here.
      *                     If used with a `$name` array, this can be set to `false` to disable attribute sorting.
+     *                     * @param string|null $before An optional string that will be prepended if the result is not empty
```
With these `$before` and `$after` arguments I'm not sure if we should really support `null`. I know we already do for the `attr()` helper, but maybe that should instead use `$before ?? ''` when calling this method. Otherwise we would cement those sloppy `null` arguments even more.

If we decide we do want to keep `null` support, that `?? ''` needs to happen inside this method to avoid PHP warnings in the future.

distantnative replied:
I don't really have an opinion here to be honest.

distantnative replied:
@bastianallgeier do you have an opinion? Guess this is the last outstanding change (plus your review then)

lukasbestle replied:
Thinking about it again, I'm fine with either solution. Only supporting `string` would be cleaner, but inconsistent. So maybe we should leave it basically like it is. With one small change: [This line here](https://github.com/getkirby/kirby/pull/4018/files#diff-9541b56c0625a5edd10a99c7249273140bb08e4b53d3cf4ce1f696034c997c78R161) should use `?? ''` to ensure that always a string is concatenated.

distantnative replied:
Let's leave it then. I would also leave the mentioned line as is to make it easier to read. `null` will always converted to an empty string in PHP. https://www.php.net/manual/en/language.types.string.php#language.types.string.casting

distantnative committed:
Helpers: Use `kirby()` instead `App::instance()`

distantnative committed:
Make `App::csrf()` non-static

distantnative committed:
Apply suggestions from code review

Co-authored-by: Lukas Bestle <lukas@getkirby.com>

distantnative committed:
Fix cs

lukasbestle committed:
Explicit `null` return value for snippet() method

Otherwise results in a `TypeError` because of the `?string` return value, which doesn't allow `void`.

lukasbestle committed:
Fix failing unit tests

lukasbestle committed:
Add unit tests for moved helper methods

lukasbestle committed:
Fix doc block indentation

distantnative committed:
Add `Cms\Html\HtmlTest`

lukasbestle committed:
Fix unneeded imports

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

lukasbestle committed:
Migrate from helpers in the core

lukasbestle committed:
`Helpers` class is now in `Cms` package

lukasbestle reviewed (commented):
Going through the use of helpers in the core, the following question just popped up:

Where do we draw the line in reuse of custom helpers in the core? Examples:

- If the `kirbytext()` helper is overridden, the `kt()` helper will automatically use it as well.
- If the `kirby()` helper is overridden, most other helpers will use the return value of this helper at the moment. But other parts of the core use `App::instance()`, which wouldn't be affected by the custom helper.

I think the first example is fine, but I'm not sure about the second example. We could either use `App::instance()` in all helpers or we could use `kirby()` throughout the whole core.

The same question for many of the other helpers. Should users actually be able to modify core behavior with custom helpers or do we want to keep the core's behavior fully predictable?

Once we have decided that, we need to check again for helper usage in the core. I stopped at the `get()` helper that's e.g. used in the Fiber callbacks a lot.

All the rest looks great!

afbora commented on config/helpers.php:
```diff
@@ -1,154 +1,136 @@
+// @codeCoverageIgnoreStart
+if (
+    defined('KIRBY_HELPER_ASSET') === false ||
+    constant('KIRBY_HELPER_ASSET') !== false
```
What do you think about moving these repeats into `Helpers` class to avoid repeats and improve readability? Something like that:

````php
if (Helpers::isHelperDefined('asset') === false) {
  // codes
}
````

### Helpers::isHelperDefined()

````php
public static function isHelperDefined(string $name) : bool 
{
    $name = 'KIRBY_HELPER_' . strtoupper($name);
    return defined($name) === true || constant($name) === false
}
````

lukasbestle replied:
Great idea! 👍

distantnative replied:
I didn't do it in the first place because I think it makes it much harder for people who face conflicts with helpers what to do. Whereas when they would see the explicit if check for the constant, it would be much more self- explanatory.

lukasbestle replied:
Hm, that's a good point. It's certainly more explicit and requires less digging. On the other hand I feel it's really much cleaner with the helper method. I'd be fine with either solution.

lukasbestle committed:
Psalm plugin to detect helper use in the core

lukasbestle committed:
Fix PHPMD warning

lukasbestle committed:
More DRY helper override checking

lukasbestle committed:
Ignore code coverage on helper override check

distantnative commented on config/areas/users/dialogs.php:
```diff
@@ -45,13 +47,16 @@
+                'password' => $kirby->request()->get('password'),
+                'language' => $kirby->request()->get('translation'),
+                'role'     => $kirby->request()->get('role')
             ]);
```
Would be great to use your "trick" here as well to define a `$data` array with just one `$kirby->request()->get()` call for all 5 keys.

lukasbestle replied:
Here the query keys don't match (`language` vs. `translation`).

distantnative commented on src/Panel/File.php:
```diff
@@ -104,14 +105,16 @@ public function dragText(string $type = null, bool $absolute = false): string
     public function dropdown(array $options = []): array
     {
+        $file    = $this->model;
+        $request = $file->kirby()->request();
```
Maybe here too


```php
$defaults = $file->kirby()->request()->get(['view', 'update', 'delete']);
```

distantnative commented on src/Panel/Page.php:
```diff
@@ -66,14 +68,16 @@ public function dragText(string $type = null): string
     public function dropdown(array $options = []): array
     {
+        $page    = $this->model;
+        $request = $page->kirby()->request();
```
```php
$defaults = $page->kirby()->request()->get(['view', 'sort', 'delete']);
```

distantnative reviewed (commented):
Looks really good, just 3 suggestions as I really liked your combined get method calls, @lukasbestle.

afbora commented on config/helpers.php:
```diff
@@ -174,782 +127,648 @@ function dump($variable, bool $echo = true): string
+    function load(array $classmap, ?string $base = null)
+    {
+        return F::loadClasses($classmap, $base);
+    }
```
```suggestion
    function load(array $classmap, ?string $base = null): void
    {
        F::loadClasses($classmap, $base);
    }
```

afbora commented on config/helpers.php:
```diff
@@ -174,782 +127,648 @@ function dump($variable, bool $echo = true): string
+     * `$kirby->markdown($text)`
+     *
+     * @param array $data
+     * @return string
```
```suggestion
     * @param string|null $text
     * @param array $data
     * @return string
```

afbora commented on config/helpers.php:
```diff
@@ -174,782 +127,648 @@ function dump($variable, bool $echo = true): string
+     * @param bool $keepTags
+     * @return string
+     */
+    function h(?string $string, bool $keepTags = false)
```
```suggestion
    function h(?string $string, bool $keepTags = false): string
```

afbora commented on config/helpers.php:
```diff
@@ -174,782 +127,648 @@ function dump($variable, bool $echo = true): string
+     * @param bool $keepTags
+     * @return string
+     */
+    function html(?string $string, bool $keepTags = false)
```
```suggestion
    function html(?string $string, bool $keepTags = false): string
```

afbora commented on config/helpers.php:
```diff
@@ -174,782 +127,648 @@ function dump($variable, bool $echo = true): string
+     * @param bool $return
+     * @return string
+     */
+    function snippet($name, $data = [], bool $return = false)
```
```suggestion
     * @return string|null
     */
    function snippet($name, $data = [], bool $return = false): ?string
```

afbora commented on config/helpers.php:
```diff
@@ -174,782 +127,648 @@ function dump($variable, bool $echo = true): string
+     * @param string|null $locale
+     * @return string
+     */
+    function tt(string $key, $fallback = null, ?array $replace = null, ?string $locale = null)
```
```suggestion
    function tt(string $key, $fallback = null, ?array $replace = null, ?string $locale = null): string
```

lukasbestle committed:
Apply suggestions from code review

afbora reviewed (approved):
Looks and works great 🎉
